src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -2831,22 +2833,30 @@ private int[] SerializeMethodBodies(BlobBuilder ilBuilder, PdbWriter nativePdbWr
 
                 if (method.HasBody())
                 {
-                    body = method.GetBody(Context);
-                    Debug.Assert(body != null || allowMissingMethodBodies);
-
-                    if (body != null)
+                    if (metadataOnly)
                     {
-                        localSignatureHandleOpt = this.SerializeLocalVariablesSignature(body);
-
-                        // TODO: consider parallelizing these (local signature tokens can be piped into IL serialization & debug info generation)
-                        bodyOffset = SerializeMethodBody(encoder, body, localSignatureHandleOpt, ref mvidStringHandle, ref mvidStringFixup);
-
-                        nativePdbWriterOpt?.SerializeDebugInfo(body, localSignatureHandleOpt, customDebugInfoWriter);
+                        bodyOffset = SerializeThrowNullMethodBody(encoder, ref mvidStringHandle, ref mvidStringFixup);
+                        localSignatureHandleOpt = default(StandaloneSignatureHandle);
+                        body = null;
                     }
                     else
                     {
-                        bodyOffset = 0;
-                        localSignatureHandleOpt = default(StandaloneSignatureHandle);
+                        body = method.GetBody(Context);
+
+                        if (body != null)
+                        {
+                            localSignatureHandleOpt = this.SerializeLocalVariablesSignature(body);
+
+                            // TODO: consider parallelizing these (local signature tokens can be piped into IL serialization & debug info generation)
+                            bodyOffset = SerializeMethodBody(encoder, body, localSignatureHandleOpt, ref mvidStringHandle, ref mvidStringFixup);
+
+                            nativePdbWriterOpt?.SerializeDebugInfo(body, localSignatureHandleOpt, customDebugInfoWriter);
+                        }
+                        else
+                        {
+                            bodyOffset = 0;
```

Comment:
Yes, I think this is an error recovery. But I don't think we should emit "throw null" however, as we do not expect PEVerify to succeed anyways.

---
In reply to: [106022809](https://github.com/dotnet/roslyn/pull/17558#discussion_r106022809) [](ancestors = 106022809)